internal/datacoord/garbage_collector.go

```diff
@@ -182,112 +193,173 @@ func (gc *garbageCollector) work() {
+		segmentID, err := storage.ParseSegmentIDByBinlog(gc.option.cli.RootPath(), chunkInfo.FilePath)
 		if err != nil {
-			log.Error("failed to list files with prefix",
-				zap.String("prefix", prefix),
-				zap.Error(err),
-			)
+			unexpectedFailure.Inc()
+			logger.Warn("garbageCollector recycleUnusedBinlogFiles parse segment id error",
+				zap.String("filePath", chunkInfo.FilePath),
+				zap.Error(err))
+			return true
 		}
-		cost := time.Since(startTs)
-		segmentMap, filesMap := getMetaMap()
-		metrics.GarbageCollectorListLatency.
-			WithLabelValues(fmt.Sprint(paramtable.GetNodeID()), labels[idx]).
-			Observe(float64(cost.Milliseconds()))
-		log.Info("gc scan finish list object", zap.String("prefix", prefix), zap.Duration("time spent", cost), zap.Int("keys", len(infoKeys)))
-		for i, infoKey := range infoKeys {
-			total++
-			_, has := filesMap[infoKey]
-			if has {
-				valid++
-				continue
-			}
 
-			segmentID, err := storage.ParseSegmentIDByBinlog(gc.option.cli.RootPath(), infoKey)
-			if err != nil {
-				missing++
-				log.Warn("parse segment id error",
-					zap.String("infoKey", infoKey),
-					zap.Error(err))
-				continue
-			}
+		segment := gc.meta.GetSegment(segmentID)
+		if checker(chunkInfo, segment) {
+			valid++
+			logger.Info("garbageCollector recycleUnusedBinlogFiles skip file since it is valid", zap.String("filePath", chunkInfo.FilePath), zap.Int64("segmentID", segmentID))
+			return true
+		}
 
-			if strings.Contains(prefix, common.SegmentInsertLogPath) &&
-				segmentMap.Contain(segmentID) {
-				valid++
-				continue
-			}
+		// ignore error since it could be cleaned up next time
+		file := chunkInfo.FilePath
+		future := gc.option.removeObjectPool.Submit(func() (struct{}, error) {
+			logger := logger.With(zap.String("file", file))
+			logger.Info("garbageCollector recycleUnusedBinlogFiles remove file...")
 
-			// not found in meta, check last modified time exceeds tolerance duration
-			if time.Since(modTimes[i]) > gc.option.missingTolerance {
-				// ignore error since it could be cleaned up next time
-				removedKeys = append(removedKeys, infoKey)
-				err = gc.option.cli.Remove(ctx, infoKey)
-				if err != nil {
-					missing++
-					log.Error("failed to remove object",
-						zap.String("infoKey", infoKey),
-						zap.Error(err))
-				}
+			if err = gc.option.cli.Remove(ctx, file); err != nil {
+				log.Warn("garbageCollector recycleUnusedBinlogFiles remove file failed", zap.Error(err))
+				unexpectedFailure.Inc()
+				return struct{}{}, err
 			}
-		}
-	}
-	metrics.GarbageCollectorRunCount.WithLabelValues(fmt.Sprint(paramtable.GetNodeID())).Add(1)
-	log.Info("scan file to do garbage collection",
+			log.Info("garbageCollector recycleUnusedBinlogFiles remove file success")
+			removed.Inc()
+			return struct{}{}, nil
+		})
+		futures = append(futures, future)
+		return true
+	})
+	// Wait for all remove tasks done.
+	// error is logged, and can be ignored here.
+	_ = conc.BlockOnAll(futures...)
```

Comment:
print the error message?

Reply:
error message is already printed at the closure.
It will repeated print the error same with that.

Reply:
error message is already printed at the closure.
It will repeated print the error same with that.